internal/args/unmarshal.go

```diff
@@ -190,70 +185,95 @@ func set(dest reflect.Value, keys []string, value string) error {
+		if err != nil {
+			return fmt.Errorf("invalid index: '%s' is not a positive integer", argNameWords[0])
 		}
 
-		// We create a new value call set and append it to the slice
-		newValue := reflect.New(dest.Type().Elem())
-		err := set(newValue.Elem(), keys, value)
-		dest.Set(reflect.Append(dest, newValue.Elem()))
-		return err
+		// Make sure array is big enough to access the correct index.
+		diff := int(index) - dest.Len()
+		switch {
+		case diff > 2:
+			return fmt.Errorf("missing indices in the array: trying to set array at index %d before indices %v", index, missingIndices(int(index), dest.Len()))
+		case diff == 1:
+			return fmt.Errorf("missing index in the array: trying to set array at index %d before index %d", index, index-1)
+		case diff == 0:
+			// Append one element to our slice.
+			dest.Set(reflect.AppendSlice(dest, reflect.MakeSlice(dest.Type(), 1, 1)))
+		case diff < 0:
+			// Element already exist at current index.
+		}
+
+		// Recursively call set without the index word
+		return set(dest.Index(int(index)), argNameWords[1:], value)
 
 	case reflect.Map:
 		// If map is nil we create it.
 		if dest.IsNil() {
 			dest.Set(reflect.MakeMap(dest.Type()))
 		}
-		if len(keys) == 0 {
+		if len(argNameWords) == 0 {
 			return fmt.Errorf("cannot handle map with no subkey, value '%v'", value)
 		}
 		// Create a new value call set and add result in the map
 		newValue := reflect.New(dest.Type().Elem())
-		err := set(newValue.Elem(), keys[1:], value)
-		dest.SetMapIndex(reflect.ValueOf(keys[0]), newValue.Elem())
+		err := set(newValue.Elem(), argNameWords[1:], value)
+		dest.SetMapIndex(reflect.ValueOf(argNameWords[0]), newValue.Elem())
 		return err
 
 	case reflect.Struct:
-		if len(keys) == 0 {
+		if len(argNameWords) == 0 {
 			return fmt.Errorf("cannot unmarshal a struct %T with not field name", dest.Interface())
 		}
 
 		// try to find the correct field in the struct.
-		fieldName := strcase.ToPublicGoName(keys[0])
+		fieldName := strcase.ToPublicGoName(argNameWords[0])
 		field := dest.FieldByName(fieldName)
 		if !field.IsValid() {
-			return fmt.Errorf("unknown argument with name %s", keys[0])
+			return fmt.Errorf("unknown argument with name %s", argNameWords[0])
 		}
 		// Set the value of the field
-		return set(field, keys[1:], value)
+		return set(field, argNameWords[1:], value)
 
-	default:
-		return fmt.Errorf("don't know how to unmarshal type %T", dest.Interface())
 	}
+	return fmt.Errorf("don't know how to unmarshal type %T", dest.Interface())
+}
+
+// missingIndices returns a string of all the missing indices between index and length.
+// e.g.: missingIndices(index=5, length=0) should return "3,2,1,0"
+// e.g.: missingIndices(index=5, length=2) should return "3,2"
+// e.g.: missingIndices(index=99999, length=0) should return "99998,99997,99996,99995,99994,99993,99992,99991,99990,..."
+func missingIndices(index, length int) string {
+	s := ""
+	for i := index - 1; i >= length; i-- {
+		if s != "" {
```

Comment:
why not computing a slice and call `strings.Join`

Reply:
It is more code